File: JavaScript/2dArrayQuestions/wordSearch.js
function exist(board, word) {
    const rows = board.length;
    const cols = board[0].length;
    function dfs(row, col, index) {
        if (index === word.length) return true;
        if (row < 0 || col < 0 || row >= rows || col >= cols || board[row][col] !== word[index]) {
            return false;
        }
        const temp = board[row][col]; 
        board[row][col] = '#'; 
        const found = dfs(row + 1, col, index + 1) ||
                      dfs(row - 1, col, index + 1) ||
                      dfs(row, col + 1, index + 1) ||
                      dfs(row, col - 1, index + 1);
        board[row][col] = temp; 
        return found;
    }
    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            if (dfs(i, j, 0)) return true;
        }
    }
    return false;
}

const board = [['A', 'B', 'C', 'E'], ['S', 'F', 'C', 'S'], ['A', 'D', 'E', 'E']];
console.log('Word search ABCCED:', exist(board, 'ABCCED')); 
console.log('Word search SEE:', exist(board, 'SEE')); 
console.log('Word search ABCB:', exist(board, 'ABCB')); 

File: src/wordSearch.java
public class wordSearch {
    public static void main(String[] args) {
        char[][] grid = {
                {'A', 'B', 'C', 'E'},
                {'S', 'F', 'C', 'S'},
                {'A', 'D', 'E', 'E'}
        };

        String word1 = "ABC";
        String word2 = "SCF";
        String word3 = "ABCB";

        System.out.println("Grid:");
        printGrid(grid);

        System.out.println("\nWord: " + word1 + ", Exists: " + exist(grid, word1));
        System.out.println("Word: " + word2 + ", Exists: " + exist(grid, word2));
        System.out.println("Word: " + word3 + ", Exists: " + exist(grid, word3));
    }

    public static boolean exist(char[][] grid, String word) {
        if (grid == null || grid.length == 0 || grid[0].length == 0 || word == null || word.length() == 0) {
            return false;
        }

        int rows = grid.length;
        int cols = grid[0].length;

        // Check horizontally
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j <= cols - word.length(); j++) {
                if (checkHorizontal(grid, i, j, word)) {
                    return true;
                }
            }
        }

        // Check vertically
        for (int j = 0; j < cols; j++) {
            for (int i = 0; i <= rows - word.length(); i++) {
                if (checkVertical(grid, i, j, word)) {
                    return true;
                }
            }
        }

        return false;
    }

    private static boolean checkHorizontal(char[][] grid, int row, int col, String word) {
        int length = word.length();
        for (int k = 0; k < length; k++) {
            if (grid[row][col + k] != word.charAt(k)) {
                return false;
            }
        }
        return true;
    }

    private static boolean checkVertical(char[][] grid, int row, int col, String word) {
        int length = word.length();
        for (int k = 0; k < length; k++) {
            if (grid[row + k][col] != word.charAt(k)) {
                return false;
            }
        }
        return true;
    }

    public static void printGrid(char[][] grid) {
        for (char[] row : grid) {
            for (char c : row) {
                System.out.print(c + " ");
            }
            System.out.println();
        }
    }
}
